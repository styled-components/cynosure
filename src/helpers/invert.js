// @flow

import parseColorString from '../internalHelpers/_parseColorString'
import rgb from './rgb'
import rgba from './rgba'

/**
 * Returns the inverted color.
 */
function invert(color: string): string {
  // parse color string to hsl
  const value = parseColorString(color)
  return value.alpha === undefined || value.alpha >= 1
    ? rgb(255 - value.red, 255 - value.green, 255 - value.blue)
    : rgba(255 - value.red, 255 - value.green, 255 - value.blue, value.alpha)
}

export default invert
