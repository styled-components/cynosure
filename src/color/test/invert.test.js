// @flow
import invert from '../invert'

describe('invert', () => {
  it('should invert a reduced hex color', () => {
    expect(invert('#448')).toMatchSnapshot()
  })

  it('should invert a hex color', () => {
    expect(invert('#CCCD64')).toMatchSnapshot()
  })

  it('should invert an 8-digit hex color', () => {
    expect(invert('#6564CDB3')).toMatchSnapshot()
  })

  it('should invert a color with opacity', () => {
    expect(invert('rgba(101,100,205,0.7)')).toMatchSnapshot()
  })
})
