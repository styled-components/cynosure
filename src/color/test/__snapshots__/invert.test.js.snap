// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invert should invert a 8-digit hex color 1`] = `"rgba(51,50,155,0.6549019607843137)"`;

exports[`invert should invert a color with opacity 1`] = `"rgba(154,155,50,0.7)"`;

exports[`invert should invert a hex color 1`] = `"#33329b"`;

exports[`invert should invert a reduced hex color 1`] = `"#bb7"`;
